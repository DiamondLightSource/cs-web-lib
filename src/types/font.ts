import { CSSProperties } from "react";

// assigned int values starting from zero
// also functions as a type
export enum FontStyle {
  Regular,
  Bold,
  Italic,
  BoldItalic
}

export class Font {
  // If no number is provided inherit from CSS.
  private size?: number;
  private style: FontStyle;
  private typeface: string;
  // TODO: name is currently unused
  private name?: string;

  public constructor(
    size?: number,
    style?: FontStyle,
    typeface?: string,
    name?: string
  ) {
    this.typeface = typeface ?? "Liberation sans";
    this.style = style ?? FontStyle.Regular;
    this.size = size;
    this.name = name;
  }

  /**
   * Returns a CSSProperties object containing all parameters defined by the
   * user and default properties for optional parameters not input by the user
   */
  public css(): CSSProperties {
    // Convert from px to REM by dividing pixels by default browser font size 16px
    // If a user has manually adjusted their browser font size, this might not be scaled
    // correctly, but given bob screens are pixel placed we shouldn't expect this to scale
    const fontSize = this.size ? `${this.size / 16}rem` : undefined;
    const fontWeight =
      this.style === FontStyle.Bold || this.style === FontStyle.BoldItalic
        ? "bold"
        : "normal";
    const fontStyle =
      this.style === FontStyle.Italic || this.style === FontStyle.BoldItalic
        ? "italic"
        : "normal";
    const style: CSSProperties = {
      // Fall back to sans-serif.
      fontFamily: `${this.typeface},sans-serif`,
      fontWeight: fontWeight,
      fontStyle: fontStyle
    };
    if (fontSize) {
      style.fontSize = fontSize;
    }
    return style;
  }
}
