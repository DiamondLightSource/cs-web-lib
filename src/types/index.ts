export { AbsolutePosition } from "./position";
export { RelativePosition } from "./position";
export { Color } from "./color";
export { PV } from "./pv";
export { Font, FontStyle } from "./font";
export { Border, BorderStyle } from "./border";
