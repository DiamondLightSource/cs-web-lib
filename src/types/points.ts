export class Point {
  public x: number;
  public y: number;

  public constructor(x: number, y: number) {
    this.x = x;
    this.y = y;
  }
}

export class Points {
  public values: Array<Point>;

  public constructor(values: Array<Point>) {
    this.values = values;
  }
}
