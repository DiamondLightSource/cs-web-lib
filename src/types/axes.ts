import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public index: number;
  public autoScale?: boolean;
  public autoScaleThreshold?: number;
  public axisColor?: Color;
  public axisTitle?: string;
  public showGrid?: boolean;
  public gridColor?: Color;
  public dashGridLine?: boolean;
  public timeFormat?: number;
  public scaleFormat?: string;
  public scaleFont?: Font;
  public titleFont?: Font;
  public visible?: boolean;
  public logScale?: boolean;
  public leftBottomSide?: boolean;
  public maximum?: number;
  public minimum?: number;
  public yAxis?: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor(idx: number) {
    this.index = idx;
    this.autoScale = true;
    this.autoScaleThreshold = 0.95;
    this.axisColor = new Color("rgb(0, 0, 0");
    this.axisTitle = "";
    this.showGrid = false;
    this.gridColor = new Color("rgb(0, 0, 0");
    this.dashGridLine = true;
    this.timeFormat = 0;
    this.scaleFormat = "";
    this.scaleFont = new Font();
    this.titleFont = new Font(10, FontStyle.Bold);
    this.visible = true;
    this.logScale = false;
    this.leftBottomSide = true;
    this.minimum = 0;
    this.maximum = 100;
    // If axis has index 1 or above, assume y axis
    this.index > 0 ? (this.yAxis = true) : (this.yAxis = false);
  }
}

export class Axes {
  public count: number;
  public axisOptions: Axis[];

  public constructor(count: number, axes: Axis[]) {
    if (count !== axes.length) {
      throw new Error(
        `Count ${count} is not equal to number of axes ${axes.length}`
      );
    }
    this.count = count;
    this.axisOptions = axes;
  }
}
