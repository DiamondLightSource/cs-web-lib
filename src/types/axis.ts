import { Color } from "./color";
import { Font, FontStyle } from "./font";

export class Axis {
  public color: Color;
  public title: string;
  public showGrid: boolean;
  public visible: boolean;
  public logScale: boolean;
  public autoscale: boolean;
  public maximum: number;
  public minimum: number;
  public titleFont: Font;
  public scaleFont: Font;
  public onRight: boolean;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property. Uses same
   * default values as csstudio.opibuilder.xygraph.
   */
  public constructor({
    xAxis = false,
    color = new Color("rgb(0, 0, 0"),
    title = "",
    showGrid = false,
    visible = true,
    logScale = false,
    autoscale = false,
    minimum = 0,
    maximum = 100,
    titleFont = new Font(FontStyle.Bold),
    scaleFont = new Font(),
    onRight = false
  } = {}) {
    this.color = color;
    this.title = title || (xAxis ? "X" : "Y");
    this.showGrid = showGrid;
    this.visible = visible;
    this.logScale = logScale;
    this.autoscale = autoscale;
    this.minimum = minimum;
    this.maximum = maximum;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.onRight = onRight;
  }
}

export type Axes = Axis[];
