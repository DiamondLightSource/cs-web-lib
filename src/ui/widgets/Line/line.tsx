import React from "react";
import { v4 as uuidv4 } from "uuid";
import { Widget } from "../widget";
import { PVWidgetPropType, PVComponent } from "../widgetProps";
import {
  InferWidgetProps,
  FloatPropOpt,
  ColorPropOpt,
  BoolPropOpt,
  PointsProp,
  MacrosPropOpt
} from "../propTypes";
import { registerWidget } from "../register";
import { Color } from "../../../types/color";
import { Point } from "../../../types/points";

const LineProps = {
  points: PointsProp,
  macros: MacrosPropOpt,
  lineWidth: FloatPropOpt,
  lineColor: ColorPropOpt,
  backgroundColor: ColorPropOpt,
  visible: BoolPropOpt,
  transparent: BoolPropOpt,
  rotationAngle: FloatPropOpt,
  arrows: FloatPropOpt,
  arrowLength: FloatPropOpt,
  fillArrow: BoolPropOpt
};

export type LineComponentProps = InferWidgetProps<typeof LineProps> &
  PVComponent;

export const LineComponent = (props: LineComponentProps): JSX.Element => {
  const {
    visible = true,
    transparent = false,
    backgroundColor = Color.fromRgba(0, 0, 255),
    rotationAngle = 0,
    lineWidth = 3,
    points,
    arrowLength = 20,
    arrows = 0,
    fillArrow = true,
    lineColor
  } = props;

  let color = (function () {
    if (lineColor) return lineColor.toString();
    else if (backgroundColor) return backgroundColor.toString();
    else return Color.fromRgba(0, 0, 255).toString();
  })();

  if (transparent) color = "transparent";
  // const color = transparent ? "transparent" : backgroundColor.toString();
  const transform = `rotation(${rotationAngle},0,0)`;

  // Each marker definition needs a unique ID or colours overlap
  const uid = uuidv4();

  // Create a marker if arrows set
  let arrowConfig = {};
  let markerConfig = <></>;
  const linePoints = points ? points.values : [];

  if (arrows) {
    switch (arrows) {
      // Arrow from
      case 1:
        arrowConfig = { markerStart: `url(#arrow${uid})` };
        // If filled, shorten line length to prevent overlap
        if (fillArrow) {
          linePoints[0] = recalculateLineLength(
            linePoints[1],
            linePoints[0],
            arrowLength
          );
        }
        break;

      // Arrow to
      case 2:
        arrowConfig = { markerEnd: `url(#arrow${uid})` };
        // If filled, shorten line length to prevent overlap
        if (fillArrow) {
          linePoints[linePoints.length - 1] = recalculateLineLength(
            linePoints[linePoints.length - 2],
            linePoints[linePoints.length - 1],
            arrowLength
          );
        }
        break;

      // Arrow both
      case 3:
        arrowConfig = {
          markerStart: `url(#arrow${uid})`,
          markerEnd: `url(#arrow${uid})`
        };
        if (fillArrow) {
          linePoints[linePoints.length - 1] = recalculateLineLength(
            linePoints[linePoints.length - 2],
            linePoints[linePoints.length - 1],
            arrowLength
          );
          linePoints[0] = recalculateLineLength(
            linePoints[1],
            linePoints[0],
            arrowLength
          );
        }
        break;
    }
    markerConfig = (
      <defs>
        <marker
          id={`arrow${uid}`}
          refX={fillArrow ? 0 : arrowLength}
          refY={`${arrowLength / 4}`}
          markerWidth={arrowLength}
          markerHeight={arrowLength / 2}
          orient="auto-start-reverse"
          markerUnits="userSpaceOnUse"
          overflow="visible"
        >
          <path
            d={`M 0 0 L ${arrowLength} ${arrowLength / 4} L 0 ${arrowLength / 2}`}
            stroke={fillArrow ? "none" : color}
            fill={fillArrow ? color : "none"}
            overflow="visible"
            strokeWidth={lineWidth}
          />
        </marker>
      </defs>
    );
  }

  let coordinates = "";
  if (points !== undefined && visible) {
    linePoints.forEach((point: Point) => {
      coordinates += `${point.x},${point.y} `;
    });
    return (
      <svg
        display={"block"}
        viewBox={`0 0 100% 100%`}
        xmlns="http://www.w3.org/2000/svg"
        overflow={"visible"}
      >
        {markerConfig}
        <polyline
          overflow={"visible"}
          stroke={color}
          strokeWidth={lineWidth}
          fill={"none"}
          transform={transform}
          points={coordinates}
          {...arrowConfig}
        />
      </svg>
    );
  } else {
    return <></>;
  }
};

/**
 * Recalculate the length of the line when a filled arrow
 * is present. This prevents the line from obscuring the
 * arrow. SVG does not include any easy way to layer an arrow
 * on top of a line without having the line show if it is thicker
 * than the arrow width, so we shorten the line to prevent this.
 * Calculate hypoteneuse of line and shorten by arrow length, then map
 * this change to the x, y coordinates
 * @param startPoint the first set of x,y coordinates of line segment
 * @param endPoint the second set of x, y coordinates of line segment
 */
function recalculateLineLength(
  startPoint: Point,
  endPoint: Point,
  arrowLen: number
): Point {
  // Determine x and y distance between coordinate sets
  let xLen = endPoint.x - startPoint.x;
  let yLen = endPoint.y - startPoint.y;
  // Calculate hypoteneuse length
  const lineLen = Math.hypot(xLen, yLen);
  // Calculate new length by subtracting arrowlength
  const newLineLen = lineLen - arrowLen;
  // If arrowLen longer than lineLen, make line short as possible
  // Ideally shouldn't be used this way as arrow should be shorter
  // Determine what fraction smaller new length is
  const frac = (arrowLen >= lineLen ? 2 : newLineLen) / lineLen;
  // Multiply lengths by fraction to get new lengths
  xLen *= frac;
  yLen *= frac;
  // Calculate new final x y coordinates
  endPoint.x = startPoint.x + Math.round(xLen);
  endPoint.y = startPoint.y + Math.round(yLen);
  // Return newly calculated final coordinates
  return endPoint;
}

const LineWidgetProps = {
  ...LineProps,
  ...PVWidgetPropType
};

export const Line = (
  props: InferWidgetProps<typeof LineWidgetProps>
): JSX.Element => <Widget baseWidget={LineComponent} {...props} />;

registerWidget(Line, LineWidgetProps, "line");
