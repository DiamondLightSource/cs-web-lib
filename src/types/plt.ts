import { Axes, Axis } from "./axis";
import { Color } from "./color";
import { Font } from "./font";
import { Trace } from "./trace";

export class Plt {
  public title: string;
  public axes: Axes;
  public pvlist: Trace[];
  public backgroundColor: Color;
  public foregroundColor: Color;
  public scroll: boolean;
  public scrollStep: number;
  public updatePeriod: number;
  public start: string;
  public end: string;
  public showGrid: boolean;
  public titleFont: Font;
  public scaleFont: Font;
  public labelFont: Font;
  public legendFont: Font;

  /**
   * Set default values for properties not yet
   * set, otherwise use set property.
   */
  public constructor({
    title = "",
    axes = [new Axis()],
    pvlist = [new Trace()],
    background = Color.WHITE,
    foreground = Color.BLACK,
    scroll = true,
    grid = false,
    scrollStep = 5,
    updatePeriod = 1,
    titleFont = new Font(),
    labelFont = new Font(),
    legendFont = new Font(),
    scaleFont = new Font(),
    start = "1 minute",
    end = "now"
  } = {}) {
    this.backgroundColor = background;
    this.foregroundColor = foreground;
    this.title = title;
    this.scroll = scroll;
    this.scrollStep = scrollStep;
    this.axes = axes;
    this.pvlist = pvlist;
    this.updatePeriod = updatePeriod;
    this.showGrid = grid;
    this.start = start;
    this.end = end;
    this.titleFont = titleFont;
    this.scaleFont = scaleFont;
    this.labelFont = labelFont;
    this.legendFont = legendFont;
  }
}
