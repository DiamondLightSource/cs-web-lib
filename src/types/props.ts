import { Color } from "./color";
import { Font } from "./font";
import { MacroMap } from "./macros";
import { WidgetActions } from "../ui/widgets/widgetActions";
import { Border } from "./border";
import { Position } from "./position";
import { PV } from "./pv";
import { Trace } from "./trace";
import { Axes, Axis } from "./axis";
import { Points } from "./points";

export type GenericProp =
  | string
  | string[]
  | boolean
  | number
  | PV
  | Color
  | Font
  | Border
  | Position
  | Rule[]
  | MacroMap
  | WidgetActions
  | OpiFile
  | Trace[]
  | Axes
  | Axis
  | Points;

export interface Expression {
  boolExp: string;
  value: string;
  convertedValue?: GenericProp;
}

interface RulePV {
  pvName: PV;
  trigger: boolean;
}

export interface Rule {
  name: string;
  prop: string;
  outExp: boolean;
  pvs: RulePV[];
  expressions: Expression[];
}

export interface OpiFile {
  path: string;
  macros: MacroMap;
  defaultProtocol: string;
}
