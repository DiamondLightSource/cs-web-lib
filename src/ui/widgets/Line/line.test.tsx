import React from "react";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import { render } from "@testing-library/react";
import { LineComponent } from "./line";
import { Color } from "../../../types/color";

const LineRenderer = (lineProps: any): ReactTestRendererJSON => {
  return renderer
    .create(<LineComponent {...lineProps} />)
    .toJSON() as ReactTestRendererJSON;
};

describe("<LineComponent />", (): void => {
  test("matches snapshot", (): void => {
    const { asFragment } = render(
      <LineComponent
        {...({
          backgroundColor: Color.fromRgba(0, 255, 255),
          width: 20,
          height: 15,
          lineWidth: 4,
          rotationAngle: 45,
          points: {
            values: [
              { x: 2, y: 20 },
              { x: 6, y: 30 },
              { x: 15, y: 4 }
            ]
          }
        } as any)}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("default properties are added to line component", (): void => {
    const lineProps = {
      width: 20,
      height: 25,
      backgroundColor: Color.fromRgba(0, 255, 255),
      points: {
        values: [
          { x: 1, y: 10 },
          { x: 15, y: 20 },
          { x: 4, y: 15 }
        ]
      }
    };

    const svg = LineRenderer(lineProps);
    expect(svg.props.viewBox).toEqual("0 0 20 25");

    const lines = svg.children as Array<ReactTestRendererJSON>;

    expect(lines[0].props.stroke).toEqual("rgba(0,255,255,255)");
    expect(lines[0].props.strokeWidth).toEqual(1);
    expect(lines[0].props.transform).toEqual("rotation(0,0,0)");
    expect(lines[0].props.points).toEqual("1,10 15,20 4,15 ");
  });

  test("props override default properties", (): void => {
    const lineProps = {
      width: 30,
      height: 20,
      lineWidth: 15,
      backgroundColor: Color.fromRgba(0, 254, 250),
      transparent: true,
      rotationAngle: 45,
      visible: true,
      points: {
        values: [
          { x: 16, y: 4 },
          { x: 25, y: 10 },
          { x: 4, y: 15 }
        ]
      },
      arrows: 3,
      arrowLength: 10
    };

    const svg = LineRenderer(lineProps);
    expect(svg.props.viewBox).toEqual("0 0 30 20");

    const lines = svg.children as Array<ReactTestRendererJSON>;
    const marker = lines[0].children as Array<ReactTestRendererJSON>;

    // Check arrowhead definitions were created
    expect(marker[0].props.markerWidth).toEqual("10");
    expect(marker[0].props.markerHeight).toEqual("10");
    expect(marker[0].props.orient).toEqual("auto-start-reverse");

    expect(lines[1].props).toHaveProperty("markerStart");
    expect(lines[1].props).toHaveProperty("markerEnd");
    expect(lines[1].props.stroke).toEqual("rgba(0,0,0,0)");
    expect(lines[1].props.strokeWidth).toEqual(15);
    expect(lines[1].props.transform).toEqual("rotation(45,0,0)");
    expect(lines[1].props.points).toEqual("16,4 25,10 4,15 ");
  });

  test("line component not created if no points to plot", (): void => {
    const lineProps = {
      width: 20,
      height: 25,
      backgroundColor: Color.fromRgba(0, 255, 255)
    };

    const svg = LineRenderer(lineProps);
    expect(svg).toBeNull();
  });
});
