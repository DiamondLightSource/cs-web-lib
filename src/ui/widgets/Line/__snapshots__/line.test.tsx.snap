// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<LineComponent /> > matches snapshot 1`] = `
<DocumentFragment>
  <svg
    display="block"
    overflow="visible"
    viewBox="0 0 100% 100%"
    xmlns="http://www.w3.org/2000/svg"
  >
    <polyline
      fill="none"
      overflow="visible"
      points="2,20 6,30 15,4 "
      stroke="rgba(0,255,255,255)"
      stroke-width="4"
      transform="rotation(45,0,0)"
    />
  </svg>
</DocumentFragment>
`;
